### ISSUE TITLE
time out and cancel requestLocation

### ISSUE THREAD

**Issue by galeen-arizonabay** [Lyokone/flutterlocation#321]
It will be nice to have a timeout feature if the user is in a basement and can not get a GPS location(could happen on android)
Also will be nice to be able to cancel a requestLocation since the user is exiting the screen fro example and this location is no longer needed

**Comment by Lyokone** [comment]
Hey ! Thanks for your suggestions, I'll be looking into that soon :)

**Comment by gitdefllo** [comment]
> Be able to cancel a requestLocation since the user is exiting the screen  

Since `location.onLocationChanged()` is a `Stream`, you could store it as `StreamSubscription` and pause / resume / cancel the location updates. 

I did this by using a bloc and switching between two states (pause and resume). When the user navigates to a different screen, send the event to pause the stream `bloc.add(PauseLocation())`. Then, use the Future callback of the Navigator to restore it: `Navigator.push(...).then((_) => bloc.add(ResumeLocation())`.

**Comment by galeen-arizonabay** [comment]
> > Be able to cancel a requestLocation since the user is exiting the screen
> 
> Since `location.onLocationChanged()` is a `Stream`, you could store it as `StreamSubscription` and pause / resume / cancel the location updates.
> 
> I did this by using a bloc and switching between two states (pause and resume). When the user navigates to a different screen, send the event to pause the stream `bloc.add(PauseLocation())`. Then, use the Future callback of the Navigator to restore it: `Navigator.push(...).then((_) => bloc.add(ResumeLocation())`.

I'm not sure this will stop the device's GPS for continue scanning and draining battery

**Comment by gitdefllo** [comment]
True. In android, to help battery performance and prevent continuing scan, you have to call `removeLocationUpdates()` to stop the service, and make another `requestLocationUpdates()` when you resume it. Adding `start()` and `stop()` methods to this library should be easy.  
However, to be aware of the changing screen as you mentioned (like exiting the screen), it's not quite the same thing. I don't know if this lib can be aware of the client side. I found that you can use [`NavigatorObserver`](https://api.flutter.dev/flutter/widgets/RouteObserver-class.html) and use methods as `didPush` and `didPopNext` to know when the widget is visible on screen and call `start()`/`stop()` to stop draining battery. But I don't find any tools to passing this awareness to this lib.

**Comment by gitdefllo** [comment]
Actually, with a little digging, it seems it's already the case in `StreamHandlerImpl`:
```
@Override
public void onListen(Object arguments, final EventSink eventsSink) {
    ...
    location.startRequestingLocation();
}

@Override
public void onCancel(Object arguments) {
    location.mFusedLocationClient.removeLocationUpdates(location.mLocationCallback);
    ...
}
```
So in order to stop requesting updates - then saving battery -, cancel the stream should do it.
Could you confirm this @Lyokone?
